tests/run-tests.py

```diff
@@ -85,6 +85,9 @@ def open(self, path, mode):
 __import__('__injected_test')
 """
 
+# Strings used in creating test basename, in the order they get applied.
+test_basename_strings = [(":", "&"), ("..", "#"), ("./", "%"), ("/", "$")]
```

Comment:
This name mangling/unmangling is pretty tricky.  These symbols are not that easy to use in the shell (eg `#` is comment, `$` is variable prefix), and often I cat/diff these files explicitly, so need to type their names.

Maybe instead, there could be a `results/_failed_tests.txt` file which simply contains each failed test, one per line?  Or more general, `results/_manifest.txt` which contains all tests that were run, and the result (pass/FAIL).

Reply:
I'm indeed not too happy with this but changing it as you say needs some extra thought; I think such _manifests.txt would need to contain the test basename (for reference, to link it back to the .out and .exp files), and the path, and then the result. Would it be ok to instead use json like a dict of `{basename: (path, result)}`? Or you'd rather have a txt with each line `basename<sep>path<sep>result` but then with a `<sep>` which would be the ASCII record separator or something like that to avoid using a character which might occur in file paths?
Also if the file contains the result explicitly, should clean-failures then delete this file, or only the lines which failed?

Reply:
I reckon it could be a `_results.json` file.  That should make it trivial to read/modify/write from Python.  (Using the name `manifest` might be too confusing with `manifest.py`...)

> Also if the file contains the result explicitly, should clean-failures then delete this file, or only the lines which failed?

Even if the file contained only the failures, it'd still need updating when re-running only these failures, because if some started to pass they'd need to be removed from the file.

Well, that could be done simply, by just always recreating the JSON file after `run-tests.py` finishes, with the list of failures.

I suggest to do the simplest thing that attains your original goal.

Might also be worth putting the `run-tests.py` command-line-arguments in the JSON.

Reply:
> I suggest to do the simplest thing that attains your original goal.

Ok. I did add command-line arguments (would also be useful to implement a 'do previous run again' command). Could add test count etc as well, just not sure how useful that is currently.

Reply:
Output will be like (this one is pretty printed):
```
{
    "args":
    {
        "target": "unix",
        "device": "/dev/ttyACM0",
        "baudrate": 115200,
        "user": "micro",
        "password": "python",
        "test_dirs":
        [
            "./basics"
        ],
        "result_dir": "C:/Users/stijn/projects/ghb/micropython/tests/results",
        "filters":
        [
            [
                "include",
                ".+"
            ]
        ],
        "write_exp": false,
        "list_tests": false,
        "emit": "bytecode",
        "heapsize": null,
        "via_mpy": false,
        "mpy_cross_flags": "-march=host",
        "keep_path": false,
        "jobs": 8,
        "files":
        [],
        "print_failures": false,
        "clean_failures": false,
        "run_failures": false
    },
    "failed_tests":
    [
        "./basics/builtin_help.py",
        "./basics/bytes_compare3.py"
    ]
}
```

Reply:
Hmm, would maybe be nicer to add the result writing in a separate function instead of just sticking everything in between the printing code. Or at least have it before the `print("{} tests performed...` line.